### ISSUE TITLE
watchContract.py core dumps when parity fills up the hard drive

### ISSUE THREAD

**Issue by tjayrush** [Great-Hill-Corporation/quickBlocks#529]
We recently changed `watchContract.py` to kill parity whenever the disc space grows larger than 90% of the total space on the hard drive. There is a problem, I think.

It appears that Parity kills itself when it runs out of hard drive space. So this line:

    Traceback (most recent call last):
    File "/Users/jrush/src.GitHub/quickBlocks/bin/watchContract.py", line 151, in <module>
        os.kill(get_parity_pid(), signal.SIGINT) #Always running at this point
    TypeError: an integer is required

fails as indicated. I tried protecting the line with something that checked if `get_parity_pid()` returned a value, but that made it obvious that this is solving the wrong problem.

My guess is that `parity` has some sort of bug that manifests itself over time. That problem is always corrected if `parity` is restarted. So, now I want the `watchContract.py` code to restart parity every 120 minutes (2 hours) or so even if the hard drive space is not growing, but----I don't want to remove the code that kills parity if the hard drive space is taken up.

In other words, every 20 seconds (or however often I run the script), it should (a) check to see how much hard drive space is left and kill (and restart it) parity if under 100GB (this is a change away from the percentage based approach), and (b) kill parity (and restart it) every two hours even if the hard drive space is not filling up.

Is this clear? If not, please let me know, and I will try to explain better.

**Comment by crodnun** [comment]
Pushed a new version of the script with the following logic:

- Parity restart time is configurable at parameter **restart_timeout**  (seconds)
- After executing the watchers round, we always check if the restart is required and update all the stuff

Let me know if it works as intended or any tweak is required, I just added a timestamp to keep track of the seconds elapsed since last restart, and do periodic restarts every X (configured value) seconds.

**Comment by tjayrush** [comment]
Sorry to say this doesn't work. I now get

Parity is already running
Restarting Parity...
Traceback (most recent call last):
  File "/Users/jrush/src.GitHub/quickBlocks/bin/watchContract.py", line 151, in <module>
    os.kill(get_parity_pid(), signal.SIGINT) #Always running at this point
TypeError: an integer is required

Whenever I run the script.  I tried really hard to debug this myself, but I don't know how to debug with python. How do you figure out what's changed?

**Comment by crodnun** [comment]
Hi Thomas, I have reviewed the same and fixed a pair of bug. Anyways a good way to debug a python script is to use this debugger: https://docs.python.org/2/library/pdb.html

It is easy, quite similar to GDB, you execute the script in debug mode this way:

**python -m pdb watchContract.py 5**

And the you can enter commands, breakpoints, etc..

As an example of a debugging session (you have all the commands at the URL), here you have some examples:

**(Pdb) n**      -------------------> To execute next line
> /home/carlos/quickBlocks/src/monitors/scripts/watchContract.py(96)<module>()
-> percentage_threshold = 10
**(Pdb) p startup_timeout**   --------------> To print a variable value
5
(Pdb) q ------------------------ For quitting

**(Pdb) b watchContract.py:50**   ----------------------------> Breakpoint at file:line number
Breakpoint 1 at /home/carlos/quickBlocks/src/monitors/scripts/watchContract.py:50

**(Pdb) r**   ------------------------- Run the program

**Comment by tjayrush** [comment]
Okay. I'll try it. By the way--part of the reason it was failing is that my hard drive is 100% full due the bug in Parity that we are trying to fix with this script. Thanks.

**Comment by tjayrush** [comment]
That worked. I had to change the percent of disk space where it fires to 1%, but it seems to work now. I will close this and re-open if I have further problems.
